File: solutions/java/collatz-conjecture/2/src/main/java/CollatzCalculator.java
import java.util.stream.IntStream;

class CollatzCalculator {

  int computeStepCount(int start) {
    if (start < 1) throw new IllegalArgumentException("Only positive integers are allowed");
    return (int) IntStream.iterate(start, n -> n > 1, n -> n % 2 == 0 ? n / 2 : 3 * n + 1).count();
  }
}

File: solutions/kotlin/collatz-conjecture/2/src/main/kotlin/CollatzCalculator.kt
object CollatzCalculator {
    fun computeStepCount(start: Int): Int {
        tailrec fun go(n: Int, acc: Int): Int =
            if (n == 1) acc else go(n.collatzStep, acc + 1)

        require(start > 0) { "Only natural numbers are allowed"}
        return go(start, 0)
    }
}

private val Int.collatzStep: Int
    get() = if (mod(2) == 0) div(2) else times(3).plus(1)